File: zigzag/zigzag.java
import java.lang.StringBuilder;

class Solution {
    public String convert(String s, int n) {
        if (n == 1) return s;
        
        StringBuilder sb = new StringBuilder();
        int len = s.length(); // length of input 
        int toTop = 0;        // number of positions from top of zig-zag
        int toBtm = 0;        // number of positions from bottom of zig-zag
        int toAdd = 0;        // index to add to get the next char
        int j = 0;
        boolean onDiag = false; // if we are on the diagonal in zig-zag
        
        // for each row
        for (int i = 0; i < n; i++) {
            
            // if top or bottom row, chars are 2(n-1) apart
            if (i == 0 || i == n - 1)
                for (j = i; j < len; j += 2*(n - 1))
                    sb.append(s.charAt(j));
            
            // iterate through the zig-zag, building the string
            else {
                toTop = i;
                toBtm = (n-1) - i;
                if (onDiag) toAdd = 2 * toTop;
                else toAdd = 2 * toBtm;
                
                for (j = i; j < len; ) {
                    int oldJ = j;
                    j += toAdd;
                    sb.append(s.charAt(oldJ));
                    if (onDiag) toAdd = 2 * toBtm;
                    else toAdd = 2 * toTop;
                    onDiag = !onDiag;
                }
                onDiag = false;
            }
        }
        
        return sb.toString();
    }
}

File: zigzag/zigzag.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        
        # annoying edge case
        if numRows == 1:
            return s
        
        rows = [[] for _i in range(numRows)]
        
        # how many chars there are in each zigzag grouping
        zigZagSize = 2 * numRows - 2

        for idx, c in enumerate(s):
            
            # we're in the 'zig' (vertical) of a group
            # correct row is just the mod of the index
            if idx % zigZagSize <= (numRows - 1):
                rows[idx % zigZagSize].append(c)
              
            # we're in the 'zag' (diagonal) of a group
            # correct row is the zigZagSize minus the mod of the index
            else:
                rows[zigZagSize - (idx % zigZagSize)].append(c)
            
            
        return ''.join([''.join(x) for x in rows])
